src/rtree.c

```diff
@@ -104,90 +148,186 @@ rtree_node_init(tsdn_t *tsdn, rtree_t *rtree, unsigned level,
 		    ATOMIC_RELAXED);
 	} else {
-		child = (rtree_elm_t *)atomic_load_p(&elm->child_or_extent,
+		leaf = (rtree_leaf_elm_t *)atomic_load_p(&elm->child,
 		    ATOMIC_ACQUIRE);
 	}
 
-	assert(!dependent || child != NULL);
-	return child;
+	assert(!dependent || leaf != NULL);
+	return leaf;
 }
 
-static rtree_elm_t *
-rtree_child_read(tsdn_t *tsdn, rtree_t *rtree, rtree_elm_t *elm, unsigned level,
-    bool dependent) {
-	rtree_elm_t *child;
+static rtree_leaf_elm_t *
+rtree_child_leaf_read(tsdn_t *tsdn, rtree_t *rtree, rtree_node_elm_t *elm,
+    unsigned level, bool dependent) {
+	rtree_leaf_elm_t *leaf;
+
+	leaf = rtree_child_leaf_tryread(elm, dependent);
+	if (!dependent && unlikely(!rtree_leaf_valid(leaf))) {
+		leaf = rtree_leaf_init(tsdn, rtree, &elm->child);
+	}
+	assert(!dependent || leaf != NULL);
+	return leaf;
+}
+
+UNUSED static rtree_node_elm_t *
+rtree_root_node_tryread(rtree_t *rtree, bool dependent) {
+	rtree_node_elm_t *node;
+	if (dependent) {
+		node = (rtree_node_elm_t *)atomic_load_p(&rtree->root,
+		    ATOMIC_RELAXED);
+	} else {
+		node = (rtree_node_elm_t *)atomic_load_p(&rtree->root,
+		    ATOMIC_ACQUIRE);
+	}
+	assert(!dependent || node != NULL);
+	return node;
+}
 
-	child = rtree_child_tryread(elm, dependent);
-	if (!dependent && unlikely(!rtree_node_valid(child))) {
-		child = rtree_node_init(tsdn, rtree, level + 1,
-		    &elm->child_or_extent);
+UNUSED static rtree_node_elm_t *
+rtree_root_node_read(tsdn_t *tsdn, rtree_t *rtree, bool dependent) {
+	rtree_node_elm_t *node = rtree_root_node_tryread(rtree, dependent);
+	if (!dependent && unlikely(!rtree_node_valid(node))) {
+		node = rtree_node_init(tsdn, rtree, 0, &rtree->root);
 	}
-	assert(!dependent || child != NULL);
-	return child;
+	assert(!dependent || node != NULL);
+	return node;
 }
 
-static rtree_elm_t *
-rtree_subtree_tryread(rtree_t *rtree, bool dependent) {
-	rtree_elm_t *subtree;
+UNUSED static rtree_leaf_elm_t *
+rtree_root_leaf_tryread(rtree_t *rtree, bool dependent) {
+	rtree_leaf_elm_t *leaf;
 	if (dependent) {
-		subtree = (rtree_elm_t *)atomic_load_p(&rtree->root,
+		leaf = (rtree_leaf_elm_t *)atomic_load_p(&rtree->root,
 		    ATOMIC_RELAXED);
 	} else {
-		subtree = (rtree_elm_t *)atomic_load_p(&rtree->root,
+		leaf = (rtree_leaf_elm_t *)atomic_load_p(&rtree->root,
 		    ATOMIC_ACQUIRE);
 	}
-	assert(!dependent || subtree != NULL);
-	return subtree;
+	assert(!dependent || leaf != NULL);
+	return leaf;
 }
 
-static rtree_elm_t *
-rtree_subtree_read(tsdn_t *tsdn, rtree_t *rtree, bool dependent) {
-	rtree_elm_t *subtree = rtree_subtree_tryread(rtree, dependent);
-	if (!dependent && unlikely(!rtree_node_valid(subtree))) {
-		subtree = rtree_node_init(tsdn, rtree, 0, &rtree->root);
+UNUSED static rtree_leaf_elm_t *
+rtree_root_leaf_read(tsdn_t *tsdn, rtree_t *rtree, bool dependent) {
+	rtree_leaf_elm_t *leaf = rtree_root_leaf_tryread(rtree, dependent);
+	if (!dependent && unlikely(!rtree_leaf_valid(leaf))) {
```

Comment:
Well, this code isn't used in practice, but it needs to be there in order to support single-level rtrees.  I tried to keep this as symmetric with the used code as possible, for simplicity/maintainability reasons.  It does work (tested in 32-bit mode), but if we were to change anything here, it might be better to simply remove single-level rtree support.  Thoughts?